README.md
simplegmaps v0.7.0
===========

simplegmaps - Add google maps to your web without knowing squat about JavaScript

##Background
I found that other scripts was too complicated to deal with and needed much scripting just to set a marker on a map. I want to separate HTML markup from JavaScript as much as possible and I want the user of the script to be able to add markers to a map without writing a single line JavaScript.

Meet simplegmaps!

##Features
* Display one or multiple markers on your map
* Add info windows to markers with custom html markup
* Display routes on your map
* No scripting necessary
* Support for traffic layers (new!)
* Support for weather layers (new!)
* Support for automatic geo location (new!)
* Support for bicycle route layer (new!)
* Support for custom marker icons (new!)
* Support for geo location on demand, by clicking a button for example (new!)
* Custom InfowWindow - Position and style your own custom infowindow.

##Browser Support
* Google Chrome
* Internet Explorer 8+
* Firefox
* Safari 6+

###[View demo](http://www.andreasnorman.com/simplegmaps)

##Installation
```
bower install simplegmaps --save
```

##Setup
```html
<!-- You'll need jquery -->
<script src="dependencies/jquery/dist/jquery.min.js"></script>
<!-- You'll also need access to google maps api -->
<script src="https://maps.googleapis.com/maps/api/js?sensor=false" type="text/javascript"></script>
<!-- and you'll need to include simplegmaps of course! -->
<script src="../src/jquery.simplegmaps.js"></script>
```

##Usage
```javascript
$('#id_of_your_div').simplegmaps();
```

###Settings and Defaults
```javascript
defaults = {
	GeoLocation: false,
  MapOptions: {
    draggable: true,
    zoom: 8,
    scrollwheel: false,
    streetViewControl: false,
    panControl: true,
    zoomControl: true,
    zoomControlOptions: {
      style: 'DEFAULT'
    }
  },
	getRouteButton: '#simplegmaps-getroute',
	getTravelMode: '#simplegmaps-travelmode',
	routeDirections: '#simplegmaps-directions',
	externalLink: '#simplegmaps-external',
	getFromAddress: '#simplegmaps-fromaddress',
	defaultTravelMode: 'DRIVING'
};

```
* `GeoLocation`: Active or deactive automatic geolocation. Default false (inactive)
* `MapOptions`: [Google Maps MapOptions](https://developers.google.com/maps/documentation/javascript/reference?csw=1#MapOptions)
* `getRouteButton`: ID of the button used to submit the route to the map
* `getTravelMode`: ID of the select element to hold the travelmode data
* `getFromAddress`: ID of the input element to hold the address to set the route start point
* `externalLink`: ID of the link element to be used when targeting a button to open up the map in a new tab. On mobile devices either Apple Maps or Google Maps app is opened instead.
* `defaultTravelMode`: The default travel mode is nothing else specified. Choose between DRIVING, WALKING or BICYCLING


###Adding markers to a map
You can use an address or latitude and longitude to position a marker on a map using the `data` attribute.
```html
<div id="simplegmap">
	<div class="map-marker" data-title="Lorem ipsum" data-latlng="55.5897407,13.012268899999981"></div>
	<div class="map-marker" data-title="Remi" data-address="Remi 145 W 53rd St, New York, NY, United States"></div>
</div>
```

###Adding a info window to a maker
You can add any html markup you wish to an info window but you need to add the class `map-infowindow` to the outer element.
```html
<div class="map-marker" data-title="Remi" data-address="Remi 145 W 53rd St, New York, NY, United States">
	<div class="map-infowindow">
	  <h2>Remi</h2>
	  <p>Remi 145 W 53rd St, New York, NY, United States</p>
	</div>
</div>
```

###Adding a map with route funtionality
```html
<div id="simplegmap">
	<div class="map-marker" data-title="Applejack Diner" data-address="Applejack Diner 1725 Broadway New York, NY 10019"></div>
</div>

<div>
	<div>
  	<label>From address</label>
   	<input type="text" id="simplegmaps-fromaddress">
  </div>
  <div>
     <label>Travel Mode</label>
     <select class="form-control" id="simplegmaps-travelmode">
      <option value="DRIVING">Driving</option>
      <option value="WALKING">Walking</option>
      <option value="BICYCLING">Bicycling</option>
    </select>
  </div>
  <div>
    <button type="submit" id="simplegmaps-getroute">Get route</button>
  </div>
</div>
```

If you change any default element ID's you'll need to set some options when you initiate the map
```javascript
$('#simplegmap').simplegmaps({
  getRouteButton: '#simplegmaps-getroute',
  getTravelMode: '#simplegmaps-travelmode',
  getFromAddress: '#simplegmaps-fromaddress'
});
```

###Adding a map with automatic geolocation
```html
<div id="simplegmap" class="google-map"></div>
```

```javascript
$('#simplegmap').simplegmaps({
  GeoLocation: true
});
```

###Adding a map with geolocation by user click
```html
<div id="simplegmap" class="google-map"></div>
<a href="#" id="geoLocationButton">Geolocate me dawg!</a>
```

```javascript
$('#simplegmap').simplegmaps();
$('#geoLocationButton').on('click', function(event) {
  event.preventDefault();
  $('#simplegmap-1').simplegmaps('setGeoLocation');
});
```

###Adding a map with different layers
Currently this plugin supports weather, traffic and bicycle layers.

In order to make weather layer work you need to include the Google Maps weather library by appending `&libraries=weather` to the path
```html
    <script src="https://maps.googleapis.com/maps/api/js?v=3.exp&sensor=false&libraries=weather"></script>
```

```html
<div id="simplegmap" class="google-map">
  <div class="map-marker" data-title="Remi" data-address="Remi 145 W 53rd St, New York, NY, United States"></div>
</div>
<a href="#" id="toggleTraffic" class="btn btn-primary">Toggle Traffic Layer</a>
<a href="#" id="toggleWeather" class="btn btn-primary">Toggle Weather Layer</a>
<a href="#" id="toggleBicycle" class="btn btn-primary">Toggle Bicycle Layer</a>

```

```javascript
$('#simplegmap').simplegmaps();

$('#toggleTraffic').on('click', function(e) {
  e.preventDefault();
  $('#simplegmap').simplegmaps('toggleTrafficLayer');
});
$('#toggleWeather').on('click', function(e) {
  e.preventDefault();
  $('#simplegmap').simplegmaps('toggleWeatherLayer');
});
$('#toggleBicycle').on('click', function(e) {
  e.preventDefault();
  $('#simplegmap').simplegmaps('toggleBicycleLayer');
});
```

###Example using snazzy maps to spice up the map style a bit
[Snazzy Maps](http://snazzymaps.com) is a nice resource to find color and style themes for Google Maps. Snazzy Maps themes are 100% compatible with simplegmaps (as it simply works out of the box with Google Maps API).

```html
<div id="simplegmap" class="google-map">
  <div class="map-marker" data-title="Remi" data-address="Remi 145 W 53rd St, New York, NY, United States"></div>
</div>
```

```javascript
$('#simplegmap').simplegmaps({
  MapOptions: {
    styles: [{"featureType":"water","stylers":[{"visibility":"on"},{"color":"#acbcc9"}]},{"featureType":"landscape","stylers":[{"color":"#f2e5d4"}]},{"featureType":"road.highway","elementType":"geometry","stylers":[{"color":"#c5c6c6"}]},{"featureType":"road.arterial","elementType":"geometry","stylers":[{"color":"#e4d7c6"}]},{"featureType":"road.local","elementType":"geometry","stylers":[{"color":"#fbfaf7"}]},{"featureType":"poi.park","elementType":"geometry","stylers":[{"color":"#c5dac6"}]},{"featureType":"administrative","stylers":[{"visibility":"on"},{"lightness":33}]},{"featureType":"road"},{"featureType":"poi.park","elementType":"labels","stylers":[{"visibility":"on"},{"lightness":20}]},{},{"featureType":"road","stylers":[{"lightness":20}]}]
  }
});
```

###Adding custom marker icons
Use the data attribute `data-icon` to specify the path to an image you wish to use as icon.
```html
<div id="simplegmap" class="google-map">
  <div class="map-marker" data-title="Remi" data-icon="img/mmazure.png" data-address="Remi 145 W 53rd St, New York, NY, United States"></div>
</div>
```

###Creating custom info windows
If you want to totally customize the position, layout and design of the info window you'll need to first wrap your map with div that has the following CSS
```css
overflow: hidden;
position: relative;
```

You can then create your info window as usual, but instead use the class name `map-custom-infowindow`

```html
<div class="simplegmap-content">
  <div id="simplegmap" class="google-map">
    <div class="map-marker" data-title="Remi" data-address="Remi 145 W 53rd St, New York, NY, United States">
      <div class="map-custom-infowindow">
        <button type="button" class="close" aria-hidden="true">&times;</button>
        <img src="img/photo.png" alt="a house" class="img-responsive">
        <h2>Remi</h2>
        <p>Remi 145 W 53rd St, New York, NY, United States</p>
      </div>
    </div>
  </div>
</div>
```

And to make it shine, add some style to your custom info window
```css
.map-custom-infowindow {
  position: absolute;
  display: block;
  bottom: 30px;
  left: 0;
  margin-left: 15%;
  margin-right: 15%;
  width: 70%;
  min-height: 120px;
  border: 1px solid #bcb8ac;
  background-color: #fff;
  color: #333;
  z-index: 10;
  padding: 10px;
  border-radius: 10px;
}
```

##changelog
####0.7.0
* Extends routing functionality to provide [DirectionsRequest Options](https://developers.google.com/maps/documentation/javascript/3.19/reference#DirectionsService)

####0.6.0
* Removed weather layer because of [Google Maps deprication](https://developers.google.com/maps/documentation/javascript/examples/layer-weather)

####0.5.0
* Added custom info windows

####0.4.0
* Added support for custom marker icons.
* Added toggle function for bicycle layer

####0.3.2
* FIX: Error occured when no markers was placed.
* FIX: MapOptions.Center did work as expected.
* FIX: Noticed that the zoom property of MapOptions was mandatory. Added fallback to default zoom setting.

####0.3.1
* FIX: Will no longer autofit when no markers has been added.

####0.3.0
* Added support for traffic layer
* Added support for weather layer
* Added support for automatic geo location
* Added support for geo location on demand (by clicking a button for example)
* Added example with snazzy maps

####0.2.0
First public release.


CHANGELOG.md
####0.7.1
* Removed console.log from the source file.

####0.7.0
* Rewrote the plugin using a different pattern

####0.5
* Added custom info windows

####0.4
* Added support for custom marker icons.
* Added toggle function for bicycle layer

####0.3.2
* FIX: Error occured when no markers was placed.
* FIX: MapOptions.Center did work as expected.
* FIX: Noticed that the zoom property of MapOptions was mandatory. Added fallback to default zoom setting.

####0.3.1
* FIX: Will no longer autofit when no markers has been added.

####0.3.0
* Added support for traffic layer
* Added support for weather layer
* Added support for automatic geo location
* Added support for geo location on demand (by clicking a button for example)
* Added example with snazzy maps

####0.2.0
First public release.


src/s.js
(function ($) {

	$.fn.simplegmaps = function (options) {
		var defaults = {};
		var settings = $.extend({}, defaults, options);

		return this.each(function () {
			$.simplegmaps = {
				Map: {},
				options: null,
				directionsService: null,
				directionsDisplay: null,
				AppleMapLink: 'http://maps.apple.com/',
				AndroidMapLink: 'http://maps.google.com/maps',
				GenericMapLink: 'http://www.google.com/maps',


				methods: {
					init: function (options) {
						console.log(this);
						$.simplegmaps.options = options;

						$.simplegmaps.directionsService = new google.maps.DirectionsService();
						$.simplegmaps.directionsDisplay = new google.maps.DirectionsRenderer({
							draggable: true
						}),
						$.simplegmaps.methods.drawMap($(this))
					},

					drawMap: function (mapEl) {
						var infoWindow = new google.maps.InfoWindow();
						var geocoder = new google.maps.Geocoder();
						var markers = [];
						var mapInData = mapEl.clone();
						var map = new google.maps.Map(mapEl[0]); // We need [0] to get the html element instead of jQery object 

						mapInData.find('div.map-marker').each(function (i) {
							if ($(this).attr('data-latlng')) {
								var marker = new google.maps.Marker({
									map: map,
									title: $(this).data('title'),
									position: parseLatLng($(this).data('latlng'))
								});
								if ($(this).has("div.map-infowindow").length > 0) {
									var infowindow = new google.maps.InfoWindow({
										content: $(this).find('div.map-infowindow').parent().html()
									});
									google.maps.event.addListener(marker, 'click', function () {
										infowindow.open(map, marker);
									});
								}

								markers.push(marker);
							} else if ($(this).attr('data-address')) {
								var currentMarkerData = $(this);
								geocoder.geocode({
									'address': $(this).data('address')
								}, function (results, status) {
									if (status == google.maps.GeocoderStatus.OK) {
										var marker = new google.maps.Marker({
											map: map,
											title: currentMarkerData.data('title'),
											position: results[0].geometry.location
										});
										if (currentMarkerData.has("div.map-infowindow").length > 0) {
											var infowindow = new google.maps.InfoWindow({
												content: currentMarkerData.find('div.map-infowindow').parent().html()
											});
											google.maps.event.addListener(marker, 'click', function () {
												infowindow.open(map, marker);
											});
										}

										markers.push(marker);
									}
								});
							}

						});

						// We need to wait for Google to locate and place all markers
						google.maps.event.addListenerOnce(map, 'idle', function () {
							zoomToFitBounds(map, markers);

							//Register map and its markers to class variable
							$.simplegmaps.Map = {
								map: map,
								markers: markers
							}

							if ($.simplegmaps.options.mapbutton) {
								bindMapLinkButton($.simplegmaps.options.mapbutton);
							}

						});
					},

					drawRoute: function (from, TravelMode) {
						var markers = $.simplegmaps.Map.markers;

						$.simplegmaps.directionsDisplay.setMap(simplegmaps.Map[targetMap].map);
						$.simplegmaps.directionsDisplay.setPanel(document.getElementById("google-map-route-directions"));

						var request = {
							origin: from,
							destination: markers[0].position,
							travelMode: google.maps.TravelMode[TravelMode]
						};
						$.simplegmaps.directionsService.route(request, function (response, status) {
							if (status == google.maps.DirectionsStatus.OK) {
								$.simplegmaps.directionsDisplay.setDirections(response);
							}
						});
					}

				},

				addMethods: function (newMethods) {
					$.simplegmaps.methods = $.extend($.simplegmaps.methods, newMethods);
				}
			};


			/*
			    PRIVATE METHODS
			  */

			// Zooms map enough to fit all markers. But not too much. Just enough to be perfect!
			// 
			// Array: markers
			function zoomToFitBounds(map, markers) {
				var bounds = new google.maps.LatLngBounds();
				for (var i = 0; i < markers.length; ++i) {
					bounds.extend(markers[i].getPosition());
				}
				if (bounds.getNorthEast().equals(bounds.getSouthWest())) {
					var extendPoint1 = new google.maps.LatLng(bounds.getNorthEast().lat() + 0.002, bounds.getNorthEast().lng() + 0.002);
					var extendPoint2 = new google.maps.LatLng(bounds.getNorthEast().lat() - 0.002, bounds.getNorthEast().lng() - 0.002);
					bounds.extend(extendPoint1);
					bounds.extend(extendPoint2);
				}
				try {
					map.fitBounds(bounds);
					map.setCenter(bounds.getCenter())
				} catch (e) {} // Let's catch this possible error and do nothing about it. Noone will ever know.
			};

			// Takes a string with latlng in this format "55.5897407,13.012268899999981" and makes it into a latlng object
			function parseLatLng(latlngString) {
				var bits = latlngString.split(/,\s*/);
				$latlng = new google.maps.LatLng(parseFloat(bits[0]), parseFloat(bits[1]));
				return $latlng;
			};

			function bindMapLinkButton(el) {
				var markerPosition = $.simplegmaps.Map.markers[0].position.toString();
				var query = '?q=' + markerPosition;
				//console.log($.simplegmaps.Map.markers[0].position.toString());

				if (kitUtils.isAndroid()) {
					$(el).attr('href', $.simplegmaps.AndroidMapLink + query);
				} else if (kitUtils.isIOS()) {
					$(el).attr('href', $.simplegmaps.AppleMapLink + query);
				} else {
					$(el).attr('href', $.simplegmaps.GenericMapLink + query);
				}
				console.log($(el));
			};
		});
	};

})(jQuery);